New base units not usable in @context in @import-ed file

Pint 0.19 introduces a regression versus 0.18.

Run the following, e.g. `python bug.py`:
```py
from pathlib import Path

import pint


# A file that defines a new base unit and uses it in a context
Path("defs0.txt").write_text("""
foo = [FOO]

@context BAR
    [FOO] -> [mass]: value / foo * 10.0 kg
@end
""")

# A file that defines a new base unit, then imports another file…
Path("defs1.txt").write_text("""
foo = [FOO]

@import defs2.txt
""")

# …that, in turn, uses it in a context
Path("defs2.txt").write_text("""
@context BAR
    [FOO] -> [mass]: value / foo * 10.0 kg
@end
""")


# Succeeds with both pint 0.18 and pint 0.19
ureg0 = pint.UnitRegistry()
ureg0.load_definitions("defs0.txt")

assert 12.0 == ureg0("1.2 foo").to("kg", "BAR").magnitude


# Succeeds with pint 0.18; fails with pint 0.19
ureg1 = pint.UnitRegistry()
ureg1.load_definitions("defs1.txt")  # ← FAILS

assert 12.0 == ureg1("1.2 foo").to("kg", "BAR").magnitude
```

The exception is:
```
Traceback (most recent call last):
  File "/home/khaeru/.local/lib/python3.9/site-packages/pint/context.py", line 323, in from_definition
    src = to_base_func(relation.src)
  File "/home/khaeru/.local/lib/python3.9/site-packages/pint/registry.py", line 770, in get_dimensionality
    return self._get_dimensionality(input_units)
  File "/home/khaeru/.local/lib/python3.9/site-packages/pint/registry.py", line 787, in _get_dimensionality
    self._get_dimensionality_recurse(input_units, 1, accumulator)
  File "/home/khaeru/.local/lib/python3.9/site-packages/pint/registry.py", line 802, in _get_dimensionality_recurse
    reg = self._dimensions[key]
KeyError: '[FOO]'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/khaeru/vc/iiasa/units/tmp/bug.py", line 39, in <module>
    ureg1.load_definitions("defs1.txt")
  File "/home/khaeru/.local/lib/python3.9/site-packages/pint/registry.py", line 650, in load_definitions
    loaderfunc(definition)
  File "/home/khaeru/.local/lib/python3.9/site-packages/pint/registry.py", line 1629, in _load_context
    self.add_context(Context.from_definition(cd, self.get_dimensionality))
  File "/home/khaeru/.local/lib/python3.9/site-packages/pint/context.py", line 331, in from_definition
    raise DefinitionSyntaxError(
  File "<string>", line 3
pint.errors.DefinitionSyntaxError: Could not add Context BAR relation on line '3'
```

Thanks for the quick response!

I ran:
```
$ pip install git+https://github.com/hgrecco/pint.git@develop
…
Successfully installed Pint-0.20.dev2+gc6c0050                                                                                                     
```

Then I tried to run the full code from which I had distilled the minimal example above to reproduce the bug. Specifically, I ran `pytest` on IAMconsortium/units, which essentially loads the definitions.txt and related files from https://github.com/IAMconsortium/units/tree/main/iam_units/data.

I now see this exception:
```
iam_units/__init__.py:18: in <module>
    registry.load_definitions(str(Path(__file__).parent / "data" / "definitions.txt"))
../../../.local/lib/python3.9/site-packages/pint/registry.py:640: in load_definitions
    for lineno, definition in parsed_files.iter_definitions():
../../../.local/lib/python3.9/site-packages/pint/parser.py:90: in iter_definitions
    yield from self._iter_definitions(*self)
../../../.local/lib/python3.9/site-packages/pint/parser.py:79: in _iter_definitions
    yield from self._iter_definitions(*pending_files)
../../../.local/lib/python3.9/site-packages/pint/parser.py:76: in _iter_definitions
    f"found {pending_files[0].path})"
E   AttributeError: 'DefinitionFile' object has no attribute 'path'
```

If it helps, replacing `pending_files[0].path` with `pending_files[0].filename.name` in the installed code from the `develop` branch gives:
```
E   ValueError: The order of the files do not match. (expected: metrics.txt, found species.txt)
```

Thus some more details about:

> definitions.txt and related files from https://github.com/IAMconsortium/units/tree/main/iam_units/data.

These files use @import recursively. The structure is roughly this, in this specific order.

- definitions.txt
  - @import emissions.txt
    - @import species.txt
    - @import metrics.txt
      - @import A.txt
      - @import B.txt
      - @import C.txt
      - etc.

I haven't closely read the new method & related code on the branch, but perhaps it is not handling this kind of structure.

--- a/CHANGES
+++ b/CHANGES
@@ -4,7 +4,8 @@ Pint Changelog
 0.20 (unreleased)
 -----------------
 
-- Nothing changed yet.
+- Provide a method to iter the definitions in the order they appear,
+  recursing the imported files. (Issue #1498)
 
 
 0.19 (2022-04-04)

--- a/pint/parser.py
+++ b/pint/parser.py
@@ -61,7 +61,33 @@ class DefinitionFile:
 class DefinitionFiles(tuple):
     """Wrapper class that allows handling a tuple containing DefinitionFile."""
 
-    pass
+    def _iter_definitions(self, next_file, *pending_files):
+        for lineno, definition in next_file.parsed_lines:
+            if isinstance(definition, ImportDefinition):
+                if not pending_files:
+                    raise ValueError(
+                        f"No more files while trying to import {definition.path}."
+                    )
+
+                if not str(pending_files[0].filename).endswith(definition.path):
+                    raise ValueError(
+                        "The order of the files do not match. "
+                        f"(expected: {definition.path}, "
+                        f"found {pending_files[0].path})"
+                    )
+
+                yield from self._iter_definitions(*pending_files)
+            else:
+                yield lineno, definition
+
+    def iter_definitions(self):
+        """Iter all definitions in the order they appear,
+        going into the included files.
+
+        Important: This assumes that the order of the imported files
+        is the one that they will appear in the definitions.
+        """
+        yield from self._iter_definitions(*self)
 
 
 def build_disk_cache_class(non_int_type: type):

--- a/pint/parser.py
+++ b/pint/parser.py
@@ -61,8 +61,19 @@ class DefinitionFile:
 class DefinitionFiles(tuple):
     """Wrapper class that allows handling a tuple containing DefinitionFile."""
 
-    def _iter_definitions(self, next_file, *pending_files):
-        for lineno, definition in next_file.parsed_lines:
+    @staticmethod
+    def _iter_definitions(
+        pending_files: list[DefinitionFile],
+    ) -> Generator[Tuple[int, Definition]]:
+        """Internal method to iterate definitions.
+
+        pending_files is a mutable list of definitions files
+        and elements are being removed as they are yielded.
+        """
+        if not pending_files:
+            return
+        current_file = pending_files.pop(0)
+        for lineno, definition in current_file.parsed_lines:
             if isinstance(definition, ImportDefinition):
                 if not pending_files:
                     raise ValueError(
@@ -73,10 +84,10 @@ class DefinitionFiles(tuple):
                     raise ValueError(
                         "The order of the files do not match. "
                         f"(expected: {definition.path}, "
-                        f"found {pending_files[0].path})"
+                        f"found {pending_files[0].filename})"
                     )
 
-                yield from self._iter_definitions(*pending_files)
+                yield from DefinitionFiles._iter_definitions(pending_files)
             else:
                 yield lineno, definition
 
@@ -87,7 +98,7 @@ class DefinitionFiles(tuple):
         Important: This assumes that the order of the imported files
         is the one that they will appear in the definitions.
         """
-        yield from self._iter_definitions(*self)
+        yield from self._iter_definitions(list(self))
 
 
 def build_disk_cache_class(non_int_type: type):

--- a/pint/registry.py
+++ b/pint/registry.py
@@ -637,17 +637,16 @@ class BaseRegistry(metaclass=RegistryMeta):
         else:
             parsed_files = parser.DefinitionFiles([p.parse_lines(file)])
 
-        for definition_file in parsed_files[::-1]:
-            for lineno, definition in definition_file.parsed_lines:
-                if definition.__class__ in p.handled_classes:
-                    continue
-                loaderfunc = loaders.get(definition.__class__, None)
-                if not loaderfunc:
-                    raise ValueError(
-                        f"No loader function defined "
-                        f"for {definition.__class__.__name__}"
-                    )
-                loaderfunc(definition)
+        for lineno, definition in parsed_files.iter_definitions():
+            if definition.__class__ in p.handled_classes:
+                continue
+            loaderfunc = loaders.get(definition.__class__, None)
+            if not loaderfunc:
+                raise ValueError(
+                    f"No loader function defined "
+                    f"for {definition.__class__.__name__}"
+                )
+            loaderfunc(definition)
 
         return parsed_files
 

--- a/pint/testsuite/test_issues.py
+++ b/pint/testsuite/test_issues.py
@@ -901,3 +901,43 @@ if np is not None:
         type_before = type(q._magnitude)
         callable(q)
         assert isinstance(q._magnitude, type_before)
+
+    def test_issue1498(tmp_path):
+        def0 = tmp_path / "def0.txt"
+        def1 = tmp_path / "def1.txt"
+        def2 = tmp_path / "def2.txt"
+
+        # A file that defines a new base unit and uses it in a context
+        def0.write_text(
+            """
+        foo = [FOO]
+
+        @context BAR
+            [FOO] -> [mass]: value / foo * 10.0 kg
+        @end
+        """
+        )
+
+        # A file that defines a new base unit, then imports another file…
+        def1.write_text(
+            f"""
+        foo = [FOO]
+
+        @import {str(def2)}
+        """
+        )
+
+        # …that, in turn, uses it in a context
+        def2.write_text(
+            """
+        @context BAR
+            [FOO] -> [mass]: value / foo * 10.0 kg
+        @end
+        """
+        )
+
+        # Succeeds with pint 0.18; fails with pint 0.19
+        ureg1 = UnitRegistry()
+        ureg1.load_definitions(def1)  # ← FAILS
+
+        assert 12.0 == ureg1("1.2 foo").to("kg", "BAR").magnitude

--- a/pint/testsuite/test_issues.py
+++ b/pint/testsuite/test_issues.py
@@ -941,3 +941,63 @@ if np is not None:
         ureg1.load_definitions(def1)  # ← FAILS
 
         assert 12.0 == ureg1("1.2 foo").to("kg", "BAR").magnitude
+
+    def test_issue1498b(tmp_path):
+        def0 = tmp_path / "def0.txt"
+        def1 = tmp_path / "def1.txt"
+        def1_1 = tmp_path / "def1_1.txt"
+        def1_2 = tmp_path / "def1_2.txt"
+        def2 = tmp_path / "def2.txt"
+
+        # A file that defines a new base unit and uses it in a context
+        def0.write_text(
+            """
+        foo = [FOO]
+
+        @context BAR
+            [FOO] -> [mass]: value / foo * 10.0 kg
+        @end
+
+        @import def1.txt
+        @import def2.txt
+        """
+        )
+
+        # A file that defines a new base unit, then imports another file…
+        def1.write_text(
+            """
+        @import def1_1.txt
+        @import def1_2.txt
+        """
+        )
+
+        def1_1.write_text(
+            """
+        @context BAR1_1
+            [FOO] -> [mass]: value / foo * 10.0 kg
+        @end
+        """
+        )
+
+        def1_2.write_text(
+            """
+        @context BAR1_2
+            [FOO] -> [mass]: value / foo * 10.0 kg
+        @end
+        """
+        )
+
+        # …that, in turn, uses it in a context
+        def2.write_text(
+            """
+        @context BAR2
+            [FOO] -> [mass]: value / foo * 10.0 kg
+        @end
+        """
+        )
+
+        # Succeeds with pint 0.18; fails with pint 0.19
+        ureg1 = UnitRegistry()
+        ureg1.load_definitions(def0)  # ← FAILS
+
+        assert 12.0 == ureg1("1.2 foo").to("kg", "BAR").magnitude

